keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,1111 @@
+Service B -> Pod B1
+```
+and at time 2, we have
+```
+Service A -> Pod A1 , Pod A2
+Service B -> Pod B1, Pod B2
+```
+
+We need to add *One* new networking rule: the fact that there is service B which can be loadbalanced to pod B2.  Any other networking rules already exist and need not be processed (this is more true for some backends then others, i.e. for IPVS or the windows kernel, which don't require rewriting of all rules every time there's a change).  
+
+KPNG provides a "DiffStore", which allows arbitrary, generic go objects to be diffed in memory by a backend.  This can be viewed at https://github.com/kubernetes-sigs/kpng/tree/master/client/diffstore.    The overall usage of this store is relatively intuitive: Write to it continuously, and only register "differences" when looking at the Diffs.  The "Reset()" function causes the second "wave" in a series of writes to take place, such that a subsequent call to see the diff at a later time will reveal differences between the first and second series of writes.  Note that the `Get` call here will write a key if empty.
+
+```
+func ExampleStore() {
+	store := NewBufferStore[string]()
+	{
+		fmt.Fprint(store.Get("a"), "hello a")
+		store.Done()
+		store.printDiff()
+	}
+	{
+		store.Reset()
+		fmt.Fprint(store.Get("a"), "hello a")
+		store.Done()
+		store.printDiff()
+	}
+```
+The entire unit test for the diffstore which is used to cache and update the network state space on the backend side, is shown in the above `/diffstore/diffstore_test.go` file. 
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a networking technology startup I want to make my own Kube Proxy but i don't want to maintain the logic of talking to the APIServer, caching its data, or caclculating an abbreviated/proxy-focused representation of the Kubernetes networking state space.  I'd like a wholesale framework I can simply plug my logic into, and re run. 
+
+#### Story 2
+
+As a Kubernetes maintainer, I don't want to have to understand the internals of a networking backend in order to simulate or write core updates to the logic of the Kube Proxy locally.
+
+#### Story 3
+
+As a Kubernetes maintainer, I'd like to add new proxies to kubernetes-sigs repositories which aren't in-tree, but are community maintained and developed/licensed according to CNCF standards
+
+#### Story 4
+
+As an end user, I'd like to be able to easily test a Kubernetes backend's networking logic without plugging it into a real Kubernetes cluster, or maybe even use it to write networking rules that aren't directly provided by the Kubernetes API.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+- sending the full-state could be resource consuming on big clusters, but it should still be O(1) to
```

Comment:
is the full state here essentially some equivalent of all k8s services endpoint details that can exist in a cluster , dowloaded to a kube-proxy implementation ? May be good to include some details on that

Reply:
yes, basically - - details are https://github.com/kubernetes-sigs/kpng/blob/master/global-state.yaml